test/runner.spec.js

```diff
@@ -433,4 +433,87 @@ describe('Runner', function () {
       });
     });
   });
+
+  describe('.run', function () {
+    describe('"test end" event', function () {
+      it('should emit "test end" event for each test', function (done) {
+        var finishedTestTitles = [];
+
+        suite.addTest(new Test('first test', noop));
+        suite.addTest(new Test('second (failed) test', function () {
+          expect(1 + 1).to.equal(3);
+        }));
+        suite.addTest(new Test('third test', noop));
+
+        runner.on('test end', function (test) {
+          finishedTestTitles.push(test.title);
+        });
+
+        runner.run(function () {
+          finishedTestTitles.should.deepEqual([
+            'first test',
+            'second (failed) test',
+            'third test'
+          ]);
+          done();
+        });
+      });
+
+      it('should have a proper duration field', function (done) {
+        var finishedTestDuration = [];
+
+        function doSomethingForSeveralMilliseconds () {
+          var result = 0;
+          for (var i = 0; i < 10000000; i++) {
+            result = i;
+          }
```

Comment:
Let's not count on the JS engine never figuring out that it could just set result to 100000000 in a single operation, skipping the loops and thus shaving off all this time; instead, use something like `var then = Date.now() + <desired milliseconds later>; while (then < Date.now()) {}` (assuming some such technique is available in all our supported environments...) to explicitly control the delay in synchronous tests and `setTimeout` to explicitly control the delay in asynchronous tests.